### ISSUE TITLE
Modify Getting Started Example to Trigger Visual Studio Bug

### ISSUE THREAD

**Issue by betanalpha** [stan-dev/pystan#400]
#### Description:

The Visual Studio bug prevents it from compiling Stan programs that end up using Eigen in the parsed C++.  The eight schools "getting started" example in the PyStan docs, https://pystan.readthedocs.io/en/latest/getting_started.html, however, doesn't use any Eigen code and so will compile fine under Visual Studio giving user the impression that their stack will be sufficient for other Stan programs.  Later compiler errors then become extremely confusing.

It would be great to change the eight schools example to use vectors instead of arrays just to tickle the Visual Studio compiler issue and inform users that they have a problem.

**Comment by ariddell** [comment]
I've lost track of the Visual Studio bug. If this fixes things, let's do it.

**Comment by ahartikainen** [comment]
Yes we could change this.
So change this from the current.

```
schools_code = """
data {
    int<lower=0> J; // number of schools
    real y[J]; // estimated treatment effects
    real<lower=0> sigma[J]; // s.e. of effect estimates
}
parameters {
    real mu;
    real<lower=0> tau;
    real eta[J];
}
transformed parameters {
    real theta[J];
    for (j in 1:J)
    theta[j] = mu + tau * eta[j];
}
model {
    eta ~ normal(0, 1);
    y ~ normal(theta, sigma);
}
"""
```

to 

```
schools_code = """
data {
    int<lower=0> J; // number of schools
    vector[J] y; // estimated treatment effects
    vector<lower=0>[J] sigma; // s.e. of effect estimates
}
parameters {
    real mu;
    real<lower=0> tau;
    vector[J] eta;
}
transformed parameters {
    vector[J] theta;
    theta = mu + tau * eta;
}
model {
    eta ~ normal(0, 1);
    y ~ normal(theta, sigma);
}
"""
```

these were replaced from the `transformed parameters`

```
    //for (j in 1:J) 
    //  theta[j] = mu + tau * eta[j];
```

**Comment by betanalpha** [comment]
This came up in a recent course where I had everyone verify ahead of
time that they could run the 8 schools model per the readthedocs 
documentation.  Everything was good for the first exercise that didn’t
use any linear algebra, but then there were mysterious compiler
errors once we moved onto the second exercise that took me a while
to recognize.  Having the first model people fit tickle the known Visual
Studio compiler bug would go a long way towards avoiding such 
issues in the future.

**Comment by ahartikainen** [comment]
Were you able to install gcc as described in the docs?

**Comment by betanalpha** [comment]
This was at a company with preexisting Windows toolkits
built on Visual Studio.  Had the getting started example
triggered the bug then we would have been able to figure
out a workaround, probably using gcc in the weeks before 
the course, but in the moment there was little we could 
due given the permissions limitations.

**Comment by ariddell** [comment]
I don't think we should have different code than RStan/CmdStan for 
8schools. Is there some alternative?

The bug is mentioned in the PyStan docs: 
http://pystan.readthedocs.io/en/latest/windows.html The first paragraph 
there is accurate "PyStan is partially supported under Windows with the 
following caveats: .."

Also, I noticed that the docs don't link to a bug report against the 
Visual Studio compiler (and/or Eigen). Assuming that the bug will 
eventually be fixed, it would be nice to be able to monitor progress.

**Comment by sakrejda** [comment]
I don't know that visual studio will ever support eigen (?).  It's
basically an issue with template support.  That said it might make sense to
have a test function in PyStan that exercise Stan a little and provides
warnings if it finds a deficient compiler.  Seems silly to install
successfully and quietly if there's an obvious problem.

**Comment by ariddell** [comment]
If we're sure that VS will not support Eigen in the next 2 years, then 
it does make sense to block installation. But I don't know of any 
example of a Python program actually doing this. We'd have to figure out 
a careful way of doing this. (also, I'd like to make sure that we're 
sure the bug will not be fixed.)

In the short term, let's state in no uncertain terms (bold in 
documentation) that PyStan does not work under Windows with VS. How does 
that sound?

**Comment by betanalpha** [comment]
We can and do state that, but it doesn’t mean that people
will read the instructions carefully.  Having a no-go example
whose compilation and running gives a very high probability
of a useful install would go a long way towards improving the
user installation experience.

**Comment by ariddell** [comment]
Ok, let's add such a thing. Is there a (short) real-world model which is 
not 8 schools which you'd recommend?

**Comment by bob-carpenter** [comment]
Can you do anything if you can't get Eigen working?  

I'd suggest just a simple linear regression with no vectors.  I never liked 8 schools since it's a measurement error/meta-analysis model, which seems awfully complicated to me for "hello world".  

```
data {
  int<lower = 0> N;
  real y[N];
  real x[N];
}
parameters {
  real<lower = 0> sigma;
  real alpha;
  real beta;
}
model {
  for (n in 1:N)
    y[n] ~ normal(alpha + beta * x[n], sigma);
  alpha ~ normal(0, 5);
  beta ~ normal(0, 3);
  sigma ~ normal(0, 2);
}
```

**Comment by betanalpha** [comment]
Any model that uses a vector anywhere will trigger the error, hence
my original suggestion to simply modify the 8 schools example to
use vectors instead of arrays.

Visual Studio will compile and run models that don’t use Eigen
anywhere, for example the model that you included below should
work fine.  But you replaced ‘real y[N]’ with ‘vector[N] y’ then it 
should fail.

**Comment by ariddell** [comment]
Ok, let's change this for 2.18.

**Comment by ahartikainen** [comment]
Just an update to here. 

When `clang-cl` gets implemented to `distutils` we can use msvc as a base and compile stuff with `clang-cl` frontend.

Until then we are moving to mingw-w64 on Windows, because msvc can't compile a working model anymore (probably due to sundials).

**Comment by ahartikainen** [comment]
Fixed by #485
